master/internal/api_experiment.go

```diff
@@ -957,67 +985,74 @@ func (a *apiServer) GetExperimentCheckpoints(
 		}
 	}
 
-	resp := &apiv1.GetExperimentCheckpointsResponse{}
-	resp.Checkpoints = []*checkpointv1.Checkpoint{}
-	switch err = a.m.db.QueryProto("get_checkpoints_for_experiment", &resp.Checkpoints, req.Id); {
-	case err == db.ErrNotFound:
-		return nil, status.Errorf(
-			codes.NotFound, "no checkpoints found for experiment %d", req.Id)
-	case err != nil:
-		return nil,
-			errors.Wrapf(err, "error fetching checkpoints for experiment %d from database", req.Id)
+	resp := &apiv1.GetExperimentCheckpointsResponse{
+		Checkpoints: []*checkpointv1.Checkpoint{},
+		Pagination:  &apiv1.Pagination{},
 	}
 
-	a.filter(&resp.Checkpoints, func(i int) bool {
-		v := resp.Checkpoints[i]
+	checkpoints := []*Checkpoint{}
+	q := db.Bun().NewSelect().
+		Model(&checkpoints).
+		Where("experiment_id = ?", req.Id)
 
-		found := false
-		for _, state := range req.States {
-			if state == v.State {
-				found = true
-				break
-			}
-		}
+	if len(req.States) > 0 {
+		states := []string{}
+		for _, s := range req.States {
+			state_text := checkpointv1.State_name[int32(s.Number())]
+			states = append(states, state_text)
```

Comment:
I think `state_text` with be a name like "STATE_ACTIVE" when the DB state is called just "ACTIVE" so this may need another layer of mapping.

Reply:
oh ok, i see in the query this is accounted for, though I think "state_unspecified" may explode? bit unfortunate, the distance between this logic.

Reply:
i think clarity-wise, this is probably the reason i prefer a query in bun right here to logic in a view.